orchestration/dagster_orchestration/hca_manage/check.py

```diff
@@ -38,12 +48,251 @@ def check_data(args: argparse.Namespace, host: str, parser: argparse.ArgumentPar
+        """
+        Determines what rows are undesired duplicates. We want to soft delete everything but the latest version for a
+        given entity_id.
+        :param target_table: The particular table to operate on.
+        :return: A set of row ids to soft delete.
+        """
+        sql_table = f"`{self.project}.datarepo_{self.dataset}.{target_table}`"
+
+        # rid -> row_id, fid -> file_id, v -> version
+
+        # allRows:          the row ids, file ids, and versions of all rows in the target table
+        # latestFids:       For all file ids that occur more than once, get the file id and the largest (latest)
+        #                   version
+        # ridsOfAllFids:    The row ids, file ids, and versions of all file ids present in latestFids; in other words,
+        #                   the (row id, file id, version) for every file id that has duplicates
+        # ridsOfLatestFids: The row ids, file ids, and versions of all file ids present in latestFids but ONLY the
+        #                   latest version rows (so a subset of ridsOfAllFids).
+        # Final query:      Get the row ids from ridsOfAllFids but exclude the row ids from ridsOfLatestFids, leaving
+        #                   us with the row ids of all non-latest version file ids. These are the rows to soft delete.
+
+        # Note: The EXCEPT DISTINCT SELECT at the end grabs all row ids of rows that AREN'T the latest version.
+        # The final subquery here is in case there are multiple rows with the same version.
+        query = f"""
+        WITH allRows AS (
+            SELECT datarepo_row_id AS rid, {target_table}_id AS fid, version AS v
+            FROM {sql_table}),
+        latestFids AS (
+            SELECT DISTINCT fid, MAX(v) AS maxv
+            FROM allRows
+            GROUP BY fid
+            HAVING COUNT(1) > 1),
+        ridsOfAllFids AS (
+            SELECT t.rid AS rid, f.fid, t.v
+            FROM latestFids f
+            JOIN allRows t
+              ON t.fid = f.fid),
+        ridsOfLatestFids AS (
+            SELECT t.rid AS rid, f.fid, f.maxv
+            FROM latestFids f
+            JOIN ridsOfAllFids t
+              ON t.fid = f.fid AND t.v = f.maxv)
+        SELECT rid
+        FROM ridsOfAllFids
+        EXCEPT DISTINCT SELECT rid
+                        FROM (
+                            SELECT MAX(rid) AS rid, fid
+                            FROM ridsOfLatestFids
+                            GROUP BY fid)
+        """
+
+        return self._hit_bigquery(query)
+
+    def get_all_table_names(self) -> set[str]:
+        """
+        Gets the table names for the target dataset.
+        :return: A set of table names.
+        """
+        query = f"""
+        SELECT table_name
+        FROM `{self.project}.datarepo_{self.dataset}.INFORMATION_SCHEMA.TABLES` WHERE table_type = "VIEW"
+        """
+
+        return self._hit_bigquery(query)
+
+    @staticmethod
```

Comment:
when i'm decomposing objects, static methods are always a signal to me that more things need to be decomposed. since they don't relate to a class's properties by definition, they're typically unrelated to the purpose of a particular class. i think `populate_row_id_csv` is a good example of that.

Reply:
So we'd pull this out into a standalone function and put it in common.py or something?

Reply:
potentially, yeah. if you find a lot of static methods that serve a common purpose, you can pull them into their own class.